src/org/rascalmpl/library/util/PathConfig.rsc

```diff
@@ -46,3 +62,298 @@ data PathConfig = pathConfig(
     list[loc] libs         = [],          
     list[Message] messages = []
 );
+
+@synopsis{Defines the parameters of mappings between qualified module names and source, target, and library files.}
+@description{
+For most languages a single `fileConfig()` instance is enough to define:
+* the mapping from source files and source folders to fully qualified module names, and back: ((srcsModule)) and ((srcsFile))
+* the mapping from binary library files to fully qualified module names and back: ((libsModule)) and ((libsFile))
+* the mapping from source files to target files in the bin folder, and back: ((binFile)) and ((binModule))
+
+Together with a ((PathConfig)) instance, the above six functions can be re-used to build a language processor that supports:
+* execution (testing) of generated files from the `bin` folder, using `libs` as run-time dependencies
+* using binary compile-time libraries, using `libs` to find binary interfaces to previously generated targets
+* packaging binary (generated) files as `jar` files to be re-used later as `libs` dependencies
+* modular language processors that work incrementally per changed source file or changed dependency
+}
+@benefits{
+* one ((fileConfig)) constant can be reused for configure all six different mapping functions.  
+* a simple `fileConfig()` constant is configured for the Rascal compiler by default (.tpl files as binary extension).
+* the mapping functions that use ((LanguageFileConfig)) can always use the same ((PathConfig)) instance for one project.
+* more complex mappings can be made by combining the six functions. For example first retrieving the module name using `srcsModule` and then 
+seeing if it exists also in one of the libraries using `libsFile`.
+}
+@pitfalls{
+* If a compiler produces multiple target files from a single source file, then you might have to configure
+different instances of ((fileConfig)) for every target `binExt`.
+* If the mapping between qualified module names and source files or binary files is different ---it has more parameters than defined by ((LanguageFileConfig))--- then you have to write your own
+versions of ((srcsModule)), ((srcsFile)), ((libsModule)), ((libsFile)), ((binFile)) and ((binModule)).
+}
+data LanguageFileConfig = fileConfig(
+    str packageSep = "::",
+    str binExt     = "tpl",
```

Comment:
Nice! I see how this works for the current Rascal compilation scheme. How can we use this for the compiler (code generation) as well? Will that have a separate `LanguageFileConfig` where `binExt = "java"` or something?

Reply:
If we decide to rewrite the Rascal checker/compiler for this, yes we'd have a fileConfig with binExt "java" or "class" depending on internal details of the compiler. However, I'm pretty sure this is not a priority to rewrite this code now. The current code is meant to be useful for DSL authors first.